pkg/logs/nats.go

```diff
@@ -0,0 +1,174 @@
+		return nil, errors.Wrap(err, "unable to connect WireGuard tunnel")
+	}
+
+	nc, err := newNatsClient(ctx, dialer, app)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating nats connection")
+	}
+
+	return &natsLogStream{nc: nc}, nil
+}
+
+// natsLogStream implements LogStream
+func (s *natsLogStream) Stream(ctx context.Context, opts *LogOptions) <-chan LogEntry {
+	out := make(chan LogEntry)
+
+	subject := fmt.Sprintf("logs.%s", opts.AppName)
+
+	if opts.RegionCode != "" {
+		subject = fmt.Sprintf("%s.%s", subject, opts.RegionCode)
+	} else {
+		subject = fmt.Sprintf("%s.%s", subject, "*")
+	}
+	if opts.VMID != "" {
+		subject = fmt.Sprintf("%s.%s", subject, opts.VMID)
+	} else {
+		subject = fmt.Sprintf("%s.%s", subject, "*")
+	}
+	// subject = fmt.Sprintf("%s.%s", subject, ">")
+
+	terminal.Debug("subscribing to nats subject: ", subject)
+
+	sub, err := s.nc.Subscribe(subject, func(msg *nats.Msg) {
+
+		var log natsLog
+
+		if err := json.Unmarshal(msg.Data, &log); err != nil {
+			terminal.Error(errors.Wrap(err, "could not parse log"))
+			return
+		}
+
+		out <- LogEntry{
+			Instance:  log.Fly.App.Instance,
+			Level:     log.Log.Level,
+			Message:   log.Message,
+			Region:    log.Fly.Region,
+			Timestamp: log.Timestamp,
+			Meta: Meta{
+				Instance: log.Fly.App.Instance,
+				Region:   log.Fly.Region,
+				Event:    struct{ Provider string }{log.Event.Provider},
+			},
+		}
+
+	})
+	if err != nil {
+		s.err = errors.Wrap(err, "could not sub to logs via nats")
+		return nil
+	}
+	go func() {
+		<-ctx.Done()
+		sub.Unsubscribe()
+	}()
+
+	return out
+}
+
+func (s *natsLogStream) Err() error {
+	return s.err
+}
+
+func newNatsClient(ctx context.Context, dialer agent.Dialer, app *api.App) (*nats.Conn, error) {
+
+	var flyConf flyConfig
+	usr, _ := user.Current()
+	flyConfFile, err := os.Open(filepath.Join(usr.HomeDir, ".fly", "config.yml"))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read fly config yml")
+	}
+	if err := yaml.NewDecoder(flyConfFile).Decode(&flyConf); err != nil {
+		return nil, errors.Wrap(err, "could not decode fly config yml")
+	}
+
+	state, ok := flyConf.WireGuardState[app.Organization.Slug]
+	if !ok {
+		return nil, errors.New("could not find org in fly config")
+	}
+
+	peerIP := state.Peer.PeerIP
+
+	var natsIPBytes [16]byte
+	copy(natsIPBytes[0:], peerIP[0:6])
+	natsIPBytes[15] = 3
+
+	natsIP := net.IP(natsIPBytes[:])
+
+	terminal.Debug("connecting to nats")
+
+	conn, err := nats.Connect(fmt.Sprintf("nats://[%s]:4223", natsIP.String()), nats.SetCustomDialer(&natsDialer{dialer, ctx}), nats.UserInfo(app.Organization.Slug, flyConf.AccessToken))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not connect to nats")
+	}
+	return conn, nil
+}
+
+type natsDialer struct {
+	agent.Dialer
+	ctx context.Context
+}
+
+func (d *natsDialer) Dial(network, address string) (net.Conn, error) {
+	return d.Dialer.DialContext(d.ctx, network, address)
+}
```

Comment:
Just curious, do you know if this works without creating a custom dialer with a context? I _think_ it would work with the normal dialer since context cancellations would happen higher up causing the cli process to end.

Reply:
I'm gonna try that

Reply:
@michaeldwan the `nats.SetCustomDialer()` option only accepts a Dialer that does not take a context.

Reply:
okay then! :)